README.md
# Tailwind-alerts

[![Latest Version on Packagist](https://img.shields.io/packagist/v/dlogon/tailwind-alerts.svg?style=flat-square)](https://packagist.org/packages/dlogon/tailwind-alerts)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/dlogon/tailwind-alerts/run-tests?label=tests)](https://github.com/dlogon/tailwind-alerts/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/dlogon/tailwind-alerts/Check%20&%20fix%20styling?label=code%20style)](https://github.com/dlogon/tailwind-alerts/actions?query=workflow%3A"Check+%26+fix+styling"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/dlogon/tailwind-alerts.svg?style=flat-square)](https://packagist.org/packages/dlogon/tailwind-alerts)

Add tailwind alerts to your page with a single facade call
```php
TailwindAlerts::addSessionMessage("This beautifull asset as been saved", TailwindAlerts::SUCCESS);
```


## Installation

You can install the package via composer:

```bash
composer require dlogon/tailwind-alerts
```

You need [tailwind](https://tailwindcss.com/) V2 or V3 and you need to check if you hace the default_alert_colors in your tailwind config file, if not, you can export the config file to change the default colors, or simply you can pass the background color you want in the level parameter.

You can publish the config file with:

```bash
php artisan vendor:publish --provider="Dlogon\TailwindAlerts\TailwindAlertsServiceProvider" --tag="config"
```

This is the contents of the published config file:

```php
"default_alert_colors" =>
    [
        "SUCCESS" => "bg-green-500",
        "ERROR" => "bg-red-500",
        "WARNING" => "bg-yellow-500",
        "INFO" => "bg-blue-500"
    ],

    "session_name" => env("TAILWIND_ALERTS", "tailwind_alerts"),

    "new_colors" => []
```

## Usage

add this [Blade template](https://laravel.com/docs/9.x/blade)   
inside your viewport html wrapper

```php
@tailwind_alerts
```
or you can choose the component style
```html
<x-tailwindalerts::tailwind-alert/>
```

Then, you can use the facade in any part of your code to add a toast message
```php
use Dlogon\TailwindAlerts\Facades\TailwindAlerts;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Route;

Route::get('/', function () {
    TailwindAlerts::addSessionMessage("This beautifull asset as been saved", TailwindAlerts::SUCCESS);
    return view('welcome');
});
```

Result
![example](example.png?raw=true "example")

the default position and type of the alert is bottom toast, you can use 4 types of alerts, calling the addSessionMessage whit the correct parameters, or calling the helper functions

Then, you can use the facade in any part of your code to add a toast message
```php
use Dlogon\TailwindAlerts\Facades\TailwindAlerts;

TailwindAlerts::addSessionMessage("Bottom toast", "bg-red-300", TailwindAlerts::BOTTOM_TOAST_CONTAINER, TailwindAlerts::TOAST_TEMPLATE );
TailwindAlerts::addBottomToastMessage("Another bottom toast", TailwindAlerts::ERROR);

TailwindAlerts::addTopToastMessage("Top toast", TailwindAlerts::WARNING);
TailwindAlerts::addHeaderMessage("Header line", TailwindAlerts::ERROR);
TailwindAlerts::addFooterMessage("Footer line", TailwindAlerts::DEFAULT_ALERT);
```
Result

![diferent alerts](https://user-images.githubusercontent.com/26014056/197311450-cdef3660-626b-43dc-9f82-18c722e08c32.png)


Where you use the component, you are able to use a javascript function to show alerts

```javascript
AlertToast.showToast("An error", AlertToast.ERROR, AlertToast.TOP_TOAST_CONTAINER, AlertToast.TOAST_TEMPLATE)
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.


## Credits

- [Diego](https://github.com/Dlogon)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

Changes to call the alerts via javascript

All notable changes to tailwind-alerts will be documented in this file.


src/TailwindAlerts.php
<?php

namespace Dlogon\TailwindAlerts;
use Illuminate\Support\Facades\Session;

class TailwindAlerts implements TailwindAlertsConsts
{
    public $session_name;
    public $alert_colors;


    public function __construct()
    {
        $this->alert_colors = \array_merge(config("tailwind-alerts.default_alert_colors"), config("tailwind-alerts.new_colors"));
        $this->session_name = config("tailwind-alerts.session_name");
    }
    public function addSessionMessage($message, $level = self::INFO, $container = self::BOTTOM_TOAST_CONTAINER, $template = self::TOAST_TEMPLATE )
    {

        $currentMessages = (array)Session::get($this->session_name);
        $levelColor = $this->alert_colors[$level] ?? $level;
        $message = [
            "level" => $levelColor,
            "message" => $message,
            "container" => $container,
            "template" => $template
        ];
        Session::forget($this->session_name);
        Session::now($this->session_name, array_merge($currentMessages, [$message]));
    }

    public function addTopToastMessage($message, $level = self::INFO)
    {
        $this->addSessionMessage($message, $level, self::TOP_TOAST_CONTAINER, self::TOAST_TEMPLATE);
    }
    public function addBottomToastMessage($message, $level = self::INFO)
    {
        $this->addSessionMessage($message, $level, self::BOTTOM_TOAST_CONTAINER, self::TOAST_TEMPLATE);
    }
    public function addHeaderMessage($message, $level = self::INFO)
    {
        $this->addSessionMessage($message, $level, self::HEADER_CONTAINER, self::LINE_TEMPLATE);
    }
    public function addFooterMessage($message, $level = self::INFO)
    {
        $this->addSessionMessage($message, $level, self::FOOTER_CONTAINER, self::LINE_TEMPLATE);
    }
}


resources/views/alerts.blade.php


<style>
    /*Toast open/load animation*/
    .alert-toast {
         -webkit-animation: slide-in-right 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
         animation: slide-in-right 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
       }

       /*Toast close animation*/
       .alert-toast input:checked~* {
         -webkit-animation: fade-out-right 0.7s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
         animation: fade-out-right 0.7s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
       }
       @-webkit-keyframes fade-out-right {
         0% {
           -webkit-transform: translateX(0);
           transform: translateX(0);
           opacity: 1
         }

         100% {
           -webkit-transform: translateX(50px);
           transform: translateX(50px);
           opacity: 0
         }
       }

       @keyframes fade-out-right {
         0% {
           -webkit-transform: translateX(0);
           transform: translateX(0);
           opacity: 1
         }

         100% {
           -webkit-transform: translateX(50px);
           transform: translateX(50px);
           opacity: 0
         }
       }
   </style>
   @if(Session::has("tailwind_alerts"))
       <div class="fixed bottom-0 right-0 m-8 w-5/6 md:w-full max-w-sm">
       @foreach (session('tailwind_alerts') as $k =>$alert)
           <div class="alert-toast" id="alert_tailwind_{{$k}}">
               <input type="checkbox" class="hidden alert_tailwind_checkbox" data-alert-id="alert_tailwind_{{$k}}" id="alert_message_buttom_{{$k}}">

               <label class="close cursor-pointer flex items-start justify-between w-full p-2 {{$alert['level']}} h-24 rounded shadow-lg text-white" title="close" for="alert_message_buttom_{{$k}}">
                   {{$alert['message']}}
               <svg class="fill-current text-white" xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 18 18">
                   <path d="M14.53 4.53l-1.06-1.06L9 7.94 4.53 3.47 3.47 4.53 7.94 9l-4.47 4.47 1.06 1.06L9 10.06l4.47 4.47 1.06-1.06L10.06 9z"></path>
               </svg>
               </label>
           </div>
       @endforeach
       </div>
   @endif

   <script>
       var elements = document.getElementsByClassName("alert_tailwind_checkbox");

       var closeDiv = function(ev) {
           var divToCloseId = ev.target.dataset.alertId;

           setTimeout(function(){
               let div = document.getElementById(divToCloseId);
               div.remove();
           },1000);
       };

       for (var i = 0; i < elements.length; i++) {
           elements[i].addEventListener('click', closeDiv, false);
       }
   </script>




src/TailwindAlertsServiceProvider.php
<?php

namespace Dlogon\TailwindAlerts;

use Illuminate\Support\ServiceProvider;
use Illuminate\Support\Facades\Blade;

class TailwindAlertsServiceProvider extends ServiceProvider
{
    public function register()
    {
        $this->app->bind('tailwindalerts', function($app) {
            return new TailwindAlerts();
        });

        $this->mergeConfigFrom(__DIR__.'/../config/tailwind-alerts.php', 'tailwind-alerts');
    }

    public function boot()
    {
        $this->loadViewsFrom(__DIR__.'/../resources/views', 'tailwindalerts');

        Blade::directive('tailwind_alerts', function ()
        {
            return \view("tailwindalerts::components/tailwind-alert");
        });
        // Blade::directive('tailwind_alerts_css', function ()
        // {
        //     return \view("tailwindalerts::styles");
        // });
        // Blade::directive('tailwind_alerts_scrips', function ()
        // {
        //     return \view("tailwindalerts::scrips");
        // });
        if ($this->app->runningInConsole())
        {

            $this->publishes([
              __DIR__.'/../config/tailwind-alerts.php' => config_path('tailwind-alerts.php'),
            ], 'config');

        }
    }

    private function css()
    {
        return <<<EOT
                /*Toast open/load animation*/
        .alert-toast {
            -webkit-animation: slide-in-right 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
            animation: slide-in-right 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
        }

        /*Toast close animation*/
        .alert-toast input:checked~* {
            -webkit-animation: fade-out-right 0.7s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
            animation: fade-out-right 0.7s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
        }

        @-webkit-keyframes fade-out-right {
            0% {
                -webkit-transform: translateX(0);
                transform: translateX(0);
                opacity: 1
            }

            100% {
                -webkit-transform: translateX(50px);
                transform: translateX(50px);
                opacity: 0
            }
        }

        @keyframes fade-out-right {
            0% {
                -webkit-transform: translateX(0);
                transform: translateX(0);
                opacity: 1
            }

            100% {
                -webkit-transform: translateX(50px);
                transform: translateX(50px);
                opacity: 0
            }
        }
        EOT;
    }
}


resources/views/components/tailwind-alert.blade.php
<div class="w-full">
    <div id="toast_template" style="display: none" class="alert-toast">
        <input type="checkbox" class="hidden alert_tailwind_checkbox">
        <label
            class="close cursor-pointer flex items-start justify-between w-full p-2 level h-24 rounded shadow-lg text-white"
            title="close">
            <p></p>
            <svg class="fill-current text-white" xmlns="http://www.w3.org/2000/svg" width="18" height="18"
                viewBox="0 0 18 18">
                <path
                    d="M14.53 4.53l-1.06-1.06L9 7.94 4.53 3.47 3.47 4.53 7.94 9l-4.47 4.47 1.06 1.06L9 10.06l4.47 4.47 1.06-1.06L10.06 9z">
                </path>
            </svg>
        </label>
    </div>
    <div id="line_template" style="display: none" class="alert-footer">
        <input type="checkbox" class="hidden alert_tailwind_checkbox">
        <div class="flex items-center justify-between w-full p-2 bg-blue-500 shadow text-white">
          <p></p>
          <label class="close cursor-pointer level" title="close">
            <svg class="fill-current text-white" xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 18 18">
              <path d="M14.53 4.53l-1.06-1.06L9 7.94 4.53 3.47 3.47 4.53 7.94 9l-4.47 4.47 1.06 1.06L9 10.06l4.47 4.47 1.06-1.06L10.06 9z"></path>
            </svg>
          </label>
        </div>
      </div>
    <div id="bottom_toast_container" class="fixed bottom-0 right-0 m-8 w-5/6 md:w-full max-w-sm">
    </div>
    <div id="top_toast_container" class="fixed top-0 right-0 m-8 w-5/6 md:w-full max-w-sm">
    </div>
    <div id="footer_container" class="w-full right-0 left-0 fixed bottom-0">
    </div>
    <div id="header_container" class="w-full right-0 left-0 fixed top-0">
    </div>

    <style>
        /*Banner open/load animation*/
        .alert-banner {
            -webkit-animation: slide-in-top 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
            animation: slide-in-top 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
        }

        /*Banner close animation*/
        .alert-banner input:checked~* {
            -webkit-animation: slide-out-top 0.5s cubic-bezier(0.550, 0.085, 0.680, 0.530) both;
            animation: slide-out-top 0.5s cubic-bezier(0.550, 0.085, 0.680, 0.530) both;
        }

        /*Footer open/load animation*/
        .alert-footer {
            -webkit-animation: slide-in-bottom 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
            animation: slide-in-bottom 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
        }

        /*Footer close animation*/
        .alert-footer input:checked~* {
            -webkit-animation: slide-out-bottom 0.5s cubic-bezier(0.550, 0.085, 0.680, 0.530) both;
            animation: slide-out-bottom 0.5s cubic-bezier(0.550, 0.085, 0.680, 0.530) both;
        }

        /*Toast open/load animation*/
        .alert-toast {
            -webkit-animation: slide-in-right 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
            animation: slide-in-right 0.5s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
        }

        /*Toast close animation*/
        .alert-toast input:checked~* {
            -webkit-animation: fade-out-right 0.7s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
            animation: fade-out-right 0.7s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
        }

        /* -------------------------------------------------------------
 * Animations generated using Animista * w: http://animista.net,
 * ---------------------------------------------------------- */

        @-webkit-keyframes slide-in-top {
            0% {
                -webkit-transform: translateY(-1000px);
                transform: translateY(-1000px);
                opacity: 0
            }

            100% {
                -webkit-transform: translateY(0);
                transform: translateY(0);
                opacity: 1
            }
        }

        @keyframes slide-in-top {
            0% {
                -webkit-transform: translateY(-1000px);
                transform: translateY(-1000px);
                opacity: 0
            }

            100% {
                -webkit-transform: translateY(0);
                transform: translateY(0);
                opacity: 1
            }
        }

        @-webkit-keyframes slide-out-top {
            0% {
                -webkit-transform: translateY(0);
                transform: translateY(0);
                opacity: 1
            }

            100% {
                -webkit-transform: translateY(-1000px);
                transform: translateY(-1000px);
                opacity: 0
            }
        }

        @keyframes slide-out-top {
            0% {
                -webkit-transform: translateY(0);
                transform: translateY(0);
                opacity: 1
            }

            100% {
                -webkit-transform: translateY(-1000px);
                transform: translateY(-1000px);
                opacity: 0
            }
        }

        @-webkit-keyframes slide-in-bottom {
            0% {
                -webkit-transform: translateY(1000px);
                transform: translateY(1000px);
                opacity: 0
            }

            100% {
                -webkit-transform: translateY(0);
                transform: translateY(0);
                opacity: 1
            }
        }

        @keyframes slide-in-bottom {
            0% {
                -webkit-transform: translateY(1000px);
                transform: translateY(1000px);
                opacity: 0
            }

            100% {
                -webkit-transform: translateY(0);
                transform: translateY(0);
                opacity: 1
            }
        }

        @-webkit-keyframes slide-out-bottom {
            0% {
                -webkit-transform: translateY(0);
                transform: translateY(0);
                opacity: 1
            }

            100% {
                -webkit-transform: translateY(1000px);
                transform: translateY(1000px);
                opacity: 0
            }
        }

        @keyframes slide-out-bottom {
            0% {
                -webkit-transform: translateY(0);
                transform: translateY(0);
                opacity: 1
            }

            100% {
                -webkit-transform: translateY(1000px);
                transform: translateY(1000px);
                opacity: 0
            }
        }

        @-webkit-keyframes slide-in-right {
            0% {
                -webkit-transform: translateX(1000px);
                transform: translateX(1000px);
                opacity: 0
            }

            100% {
                -webkit-transform: translateX(0);
                transform: translateX(0);
                opacity: 1
            }
        }

        @keyframes slide-in-right {
            0% {
                -webkit-transform: translateX(1000px);
                transform: translateX(1000px);
                opacity: 0
            }

            100% {
                -webkit-transform: translateX(0);
                transform: translateX(0);
                opacity: 1
            }
        }

        @-webkit-keyframes fade-out-right {
            0% {
                -webkit-transform: translateX(0);
                transform: translateX(0);
                opacity: 1
            }

            100% {
                -webkit-transform: translateX(50px);
                transform: translateX(50px);
                opacity: 0
            }
        }

        @keyframes fade-out-right {
            0% {
                -webkit-transform: translateX(0);
                transform: translateX(0);
                opacity: 1
            }

            100% {
                -webkit-transform: translateX(50px);
                transform: translateX(50px);
                opacity: 0
            }
        }
    </style>
    <script>
        var closeDiv = function(ev) {
            var divToCloseId = ev.target.dataset.alertId;

            setTimeout(function() {
                let div = document.getElementById(divToCloseId);
                div.remove();
            }, 1000);
        };
        var AlertToast = {
            BOTTOM_TOAST_CONTAINER: "bottom_toast_container",
            TOP_TOAST_CONTAINER: "top_toast_container",
            FOOTER_CONTAINER: "footer_container",
            HEADER_CONTAINER: "header_container",
            TOAST_TEMPLATE:"toast_template",
            LINE_TEMPLATE: "line_template",
            SUCCESS: "{{ config('tailwind-alerts.default_alert_colors.SUCCESS') }}",
            ERROR: "{{ config('tailwind-alerts.default_alert_colors.ERROR') }}",
            WARNING: "{{ config('tailwind-alerts.default_alert_colors.WARNING') }}",
            INFO: "{{ config('tailwind-alerts.default_alert_colors.INFO') }}",
            showToast: function(message, level, container = this.BOTTOM_TOAST_CONTAINER, template = this.TOAST_TEMPLATE) {

                let divId = (Math.random() + 1).toString(36).substring(7);
                let checkId = (Math.random() + 1).toString(36).substring(7);

                let div = document.getElementById(template).cloneNode(true);
                let alertContainer = document.getElementById(container);

                div.id = divId;
                let containerLabel = div.querySelector("label");
                containerLabel.setAttribute("for", checkId);

                let check = div.querySelector("input");
                check.dataset.alertId = divId;
                check.setAttribute("id", checkId);
                check.addEventListener('click', closeDiv, false);

                containerLabel.className = containerLabel.className.replace("level", level);
                div.querySelector("p").textContent = message;

                alertContainer.appendChild(div);
                div.style.display = "block";
            }
        }
    </script>
    @if (Session::has(config('tailwind-alerts.session_name')))
        @foreach (session(config('tailwind-alerts.session_name')) as $k => $alert)
            <script>
                AlertToast.showToast("{{ $alert['message'] }}", "{{ $alert['level'] }}", "{{ $alert['container'] }}", "{{ $alert['template'] }}")
            </script>
        @endforeach
    @endif

</div>
